docs/index.rst

```diff
@@ -1,260 +1,427 @@
-
-    config = confit.Configuration('myapp')
-    parser = argparse.ArgumentParser()
-    parser.add_argument('--foo', help='a parameter')
-    args = parser.parse_args()
-    config.set_args(args)
-    print(config['foo'].get())
-
-This will allow the user to override the configured value for key
-``foo`` by passing ``--foo <something>`` on the command line.
-
-.. _argparse: http://docs.python.org/dev/library/argparse.html
-.. _parse_args: http://docs.python.org/library/argparse.html#the-parse-args-method
-.. _optparse: http://docs.python.org/library/optparse.html
-
-Note that, while you can use the full power of your favorite
-command-line parsing library, you'll probably want to avoid specifying
-defaults in your argparse or optparse setup. This way, Confit can use
-other configuration sources---possibly your
-``config_default.yaml``---to fill in values for unspecified
-command-line switches. Otherwise, the argparse/optparse default value
-will hide options configured elsewhere.
-
-
-Search Paths
-------------
-
-Confit looks in a number of locations for your application's
-configurations. The locations are determined by the platform. For each
-platform, Confit has a list of directories in which it looks for a
-directory named after the application. For example, the first search
-location on Unix-y systems is ``$XDG_CONFIG_HOME/AppName`` for an
-application called ``AppName``.
-
-Here are the default search paths for each platform:
-
-* OS X: ``~/.config/app`` and ``~/Library/Application Support/app``
-* Other Unix: ``$XDG_CONFIG_HOME/app`` and ``~/.config/app``
-* Windows: ``%APPDATA%\app`` where the `APPDATA` environment variable falls
-  back to ``%HOME%\AppData\Roaming`` if undefined
-
-Users can also add an override configuration directory with an
-environment variable. The variable name is the application name in
-capitals with "DIR" appended: for an application named ``AppName``, the
-environment variable is ``APPNAMEDIR``.
-
-
-Your Application Directory
---------------------------
+Configurations can be composed with the addition operator::
+
+  config = defaults + overrides
+  config += more_overrides
+
+The result of this will be a new configuration with a union of the variables
+in ``overrides`` and ``defaults`` with the values and origins from
+``overrides`` taking precedence.
+
+``defaults`` and ``overrides`` themselves can be compositions, creating a
+tree of configurations. Generally, due to the left associativity of the ``+``
+operator in Python, only ``defaults`` will be a composition, creating a
+sequence. Summing an iterable of configurations will compose them from left
+to right, in order of lowest precedence to highest precedence.
+
+Variables have a third method that becomes useful in the context of
+compositions. ``stack`` will return a list, in precedence order starting with
+most preferred, of all the variable definitions in the configuration tree::
+
+  assert config['var'].stack()[0].get() == config['var'].get()
```

Comment:
`stack()` will be part of the general `Variable` interface. The flat variable class will have a trivial implementation, `return [self]`. Also, it will need to return an iterable, not a list, to have laziness.